Resource update mechanism based on Discover flow and UnifiedResources UI

flyinghermit opened the pull request:
- The action button for SAML apps is updated with `ButtonWithMenu` button. The menu is configured with edit menu button, which opens update dialogue, passing identifier of SAMl application which is to be updated. 

- Discover flow is updated to accommodate update flow.
A new props `updateFlow` have been added of type `DiscoverUpdateProps`:

```
export type DiscoverUpdateProps = {
  // agentMeta specifies ResourceSpec which should be used to
  // start a Discover flow.
  resourceSpec: ResourceSpec;
  // agentMeta includes data that will be used to prepopulate input fields
  // in the respective Discover compnents.
  agentMeta: AgentMeta;
};
```

Implemented to support SAML application update flow - https://github.com/gravitational/teleport.e/issues/4458

flyinghermit committed:
update SAML app login ButtonBorder button ButtonWithMenu button

flyinghermit committed:
update ResourceActionButton form SAML app to handle update flow

flyinghermit committed:
feat: update Discover flow to support update resource mechanism

flyinghermit committed:
Merge branch 'master' into sshah/saml-app-edit

flyinghermit committed:
add SamlIdpServiceProvider type to SamlMeta

flyinghermit committed:
make SamlMeta fields optional

flyinghermit committed:
prop types and update fix action button for story

gzdunek commented on web/packages/teleport/src/Discover/useDiscover.tsx:
```diff
@@ -90,11 +94,22 @@ type CustomEventInput = {
   discoveryConfigMethod?: DiscoverDiscoveryConfigMethod;
 };
 
+export type DiscoverUpdateProps = {
+  // resourceSpecForUpdate specifies ResourceSpec which should be used to
+  // start a Discover flow.
+  resourceSpec: ResourceSpec;
+  // agentMetaForUpdate includes data that will be used to prepopulate input fields
+  // in the respective Discover compnents.
+  agentMeta: AgentMeta;
+};
```
The field names in comments don't match the actual field names.

gzdunek commented on web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:
```diff
@@ -177,19 +206,32 @@ const AppLaunch = ({ app }: { app: App }) => {
     );
   }
+  function handleSamlAppEditButtonClick() {
+    setUpdateDialogOpen(true);
+    setResourceSpec({
+      name: name,
+      event: DiscoverEventResource.SamlApplication,
+      kind: ResourceKind.SamlApplication,
+      samlMeta: { preset: samlAppPreset },
+      icon: 'Application',
+      keywords: 'saml',
+    });
```
Hmm, why do we need two callbacks for that? Doesn't setting resource spec mean that the dialog should be open?
I'd call it `onUpdateResource: (resourceSpec: ResourceSpec) => void`.

flyinghermit replied:
https://github.com/gravitational/teleport/pull/43871/commits/d7c6d0d59d77102a981c11e625dec7ac6aa0d0c5#diff-ffc6bf1267af2b53c1dad2ff62eab6a4b6ae53e7b80ce49e33b62fc5b1808e5e

gzdunek commented on web/packages/teleport/src/Discover/useDiscover.tsx:
```diff
@@ -455,6 +479,7 @@ export function DiscoverProvider({
     emitErrorEvent,
     emitEvent,
     eventState,
+    isUpdateFlow: Boolean(updateFlow),
```
```suggestion
    isUpdateFlow: !!updateFlow,
```

gzdunek commented on web/packages/teleport/src/Discover/Discover.test.tsx:
```diff
@@ -204,3 +208,91 @@ describe('location state', () => {
+
+  renderUpdate({ resourceSpec: resourceSpec, agentMeta: { resourceName: '' } });
+
+  expect(screen.queryAllByTestId(ResourceKind.Server).length).toBeFalsy();
```
This seems to be easier to understand:
```suggestion
  expect(screen.queryByTestId(ResourceKind.Server)).not.toBeInTheDocument();
```
(same for other assertions)

gzdunek commented on web/packages/teleport/src/Discover/useDiscover.tsx:
```diff
@@ -231,6 +247,14 @@ export function DiscoverProvider({
   }
 
+  // trigger update Discover flow.
+  useEffect(() => {
+    if (updateFlow && updateFlow.agentMeta) {
+      updateAgentMeta(updateFlow.agentMeta);
+      onSelectResource(updateFlow.resourceSpec);
+    }
+  }, [updateFlow]);
```
This is a react anti pattern https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes.
Maybe we could replace it with setting initial value in `setState`s?

flyinghermit replied:
I am not sure that suggestion fits the bill here.
`onSelectResource` is not a react `SetStateAction` but  a function that starts discover flow. It, among setting up resource view, also dispatches Event, which is an outgoing network request.  Only `updateAgentMeta` directly wraps a SetStateAction.

gzdunek commented on web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:
```diff
@@ -177,19 +206,32 @@ const AppLaunch = ({ app }: { app: App }) => {
         rel="noreferrer"
         textTransform="none"
+        forwardedAs="a"
+        title="Log in to this SAML application"
```
I'm not sure if `this` is needed.
```suggestion
        title="Log in to SAML application"
```

flyinghermit committed:
review suggestions:
- update SamlMeta fields
- update Discover test to query single node instead of all possible resources
- remove update dialog open setAction props
- udpate comment

gzdunek commented on web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx:
```diff
@@ -177,19 +190,31 @@ const AppLaunch = ({ app }: { app: App }) => {
     return (
-      <ButtonBorder
-        as="a"
+      <ButtonWithMenu
```
Shouldn't we show `ButtonBorder` when `setResourceSpec` is empty?
Now when I run OSS, I see an edit button on a SAML app that does nothing.

flyinghermit replied:
Now returns `ButtonBorder` if setResourceSpec is empty https://github.com/gravitational/teleport/pull/43871/commits/18a5608fafa9e8423574eaf1537ad39905a9f6fc

flyinghermit committed:
return ButtonWithMenu only if setResourceSpec is provided

flyinghermit committed:
update AttributeMapping type field and comment

flyinghermit commented:
I slipped https://github.com/gravitational/teleport/pull/43871/commits/0dc5cfbbf3ecf6f12e3b6d4aa3d565a5ad2e7465 after approval, which is needed for the `e` PR https://github.com/gravitational/teleport.e/pull/4584.
